### PR TITLE
Fix methods for recalculating waypoints coordinates, names, indices, approaches.

### PR DESCRIPTION
**Pull request by Lebedinsky**
## Description

Fix methods for recalculating waypoint coordinates, indices, names, approaches [#2045 ](https://github.com/mapbox/mapbox-navigation-android/issues/2045))

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

`RouteUtils` methods for recalculating waypoints are working incorrectly when `offroad` event happened. The goal here is to fix this methods and add tests to cover all main cases. 

### Implementation

Before we calculating remaining waypoints based on waypoint index and remaining waypoints. `RouteProgress#remainingWaypointsCount()` method return number of remaining `legs` but not remaining `coordinates`. So when we calculate remaining waypoints we need get correspond coordinates based on remaining `waypoint indexes` (remaining legs) and then perform all calculations.

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/navigationui/WaypointNavigationActivity.java [hunk 1]
```diff
@@ -190,8 +192,11 @@ private void fetchRoute(Point origin, Point destination) {
       .build()
       .getRoute(new SimplifiedCallback() {
         @Override
-        public void onResponse(Call<DirectionsResponse> call, Response<DirectionsResponse> response) {
-          startNavigation(response.body().routes().get(0));
+        public void onResponse(@NotNull Call<DirectionsResponse> call, @NotNull Response<DirectionsResponse> response) {
+          DirectionsResponse directionsResponse = response.body();
+          if (directionsResponse != null) {
+            startNavigation(directionsResponse.routes().get(0));
```
Comment (andrlee) [inline app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/navigationui/WaypointNavigationActivity.java]: wouldn't `get(0)` call on an empty list result in `IndexOutOfBoundsException`?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteCallback.java [hunk 2]
```diff
@@ -29,8 +29,10 @@
   public void onResponse(@NonNull Call<DirectionsResponse> call, @NonNull Response<DirectionsResponse> response) {
     callback.onResponse(call, response);
     if (isValid(response)) {
-      String uuid = response.body().uuid();
-      sendEventWith(listener.getTime(), uuid);
+      DirectionsResponse directionsResponse = response.body();
+      if (directionsResponse != null) {
+        sendEventWith(listener.getTime(), directionsResponse.uuid());
```
Comment (andrlee) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteCallback.java]: wouldn't `directionsResponse.uuid()` returning `null` be problematic here?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 3]
```diff
@@ -65,19 +70,73 @@ public boolean isLastLeg(RouteProgress routeProgress) {
    */
   @Nullable
   public List<Point> calculateRemainingWaypoints(RouteProgress routeProgress) {
```
Comment (andrlee) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: wouldn't it be more user friendly if we'd return empty list? wouldn't need to handle null down the road.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 4]
```diff
@@ -65,19 +70,73 @@ public boolean isLastLeg(RouteProgress routeProgress) {
...
+   * Remaining waypoint indices are recalculated based on count of already achieved waypoints.
+   * <p>
+   * If the waypoint indices are empty, this method will return null.
+   *
+   * @param routeProgress for route waypoint indices and remaining coordinates
+   * @return Integer array including the origin waypoint index and the recalculated remaining ones
+   * @since 0.43.0
+   */
+  @Nullable
+  public Integer[] calculateRemainingWaypointIndices(RouteProgress routeProgress) {
```
Comment (andrlee) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: Same here, seems like returning empty array might be a better strategy as opposed to returning `null`?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 5]
```diff
@@ -94,16 +153,52 @@ public boolean isLastLeg(RouteProgress routeProgress) {
...
-    String allWaypointNames = routeOptions.waypointNames();
-    String[] names = allWaypointNames.split(SEMICOLON);
-    int coordinatesSize = routeProgress.directionsRoute().routeOptions().coordinates().size();
-    String[] remainingWaypointNames = Arrays.copyOfRange(names,
-      coordinatesSize - routeProgress.remainingWaypoints(), coordinatesSize);
-    String[] waypointNames = new String[remainingWaypointNames.length + ORIGIN_WAYPOINT_NAME_THRESHOLD];
-    waypointNames[ORIGIN_WAYPOINT_NAME] = names[ORIGIN_WAYPOINT_NAME];
-    System.arraycopy(remainingWaypointNames, FIRST_POSITION, waypointNames, SECOND_POSITION,
+    String waypointNames = routeOptions.waypointNames();
+    if (waypointNames == null) {
```
Comment (andrlee) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: empty array?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 6]
```diff
@@ -94,16 +153,52 @@ public boolean isLastLeg(RouteProgress routeProgress) {
...
+   * along the given route based on route option approaches and the progress remaining approaches.
+   * <p>
+   * If the approaches are empty, this method will return null.
+   *
+   * @param routeProgress for route approaches and remaining coordinates
+   * @return String array including the origin approach and the remaining ones
+   * @since 0.19.0
+   */
+  @Nullable
+  public String[] calculateRemainingApproaches(RouteProgress routeProgress) {
```
Comment (andrlee) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: empty array?

**Comment by Guardiola31337** [comment]
BTW @Lebedinsky some of the comments were already pointed out by @andrlee 😬 

Sorry for the inconvenience, it seems we were reviewing the PR at the same time / in parallel 😅

File: app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/navigationui/WaypointNavigationActivity.java (hunk 1 above)
Comment (Lebedinsky) [reply app/src/main/java/com/mapbox/services/android/navigation/testapp/activity/navigationui/WaypointNavigationActivity.java]: Done

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteCallback.java [hunk 7]
```diff
@@ -29,8 +29,10 @@
   public void onResponse(@NonNull Call<DirectionsResponse> call, @NonNull Response<DirectionsResponse> response) {
     callback.onResponse(call, response);
     if (isValid(response)) {
-      String uuid = response.body().uuid();
-      sendEventWith(listener.getTime(), uuid);
+      DirectionsResponse directionsResponse = response.body();
+      if (directionsResponse != null) {
```
[the remark this answers is not in the document]
Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteCallback.java]: Done

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteCallback.java (hunk 2 above)
Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationRouteCallback.java]: No, I've checked all calls sequence and in a result this uuid converts to String and sending as some metric event. So I think it is ok here.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 8]
```diff
@@ -35,7 +40,7 @@
    */
   public boolean isArrivalEvent(@NonNull RouteProgress routeProgress) {
     RouteProgressState currentState = routeProgress.currentState();
-    return currentState != null && currentState == RouteProgressState.ROUTE_ARRIVED;
+    return currentState == RouteProgressState.ROUTE_ARRIVED;
```
[the remark this answers is not in the document]
Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: Condition `currentState != null` is covered by subsequent condition `currentState == RouteProgressState.ROUTE_ARRIVED`
So it no needed here.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 9]
```diff
@@ -53,7 +58,7 @@ public boolean isLastLeg(RouteProgress routeProgress) {
   }
 
   /**
-   * Given a {@link RouteProgress}, this method will calculate the remaining coordinates
+   * Given a {@link RouteProgress}, this method will calculate the remaining waypoints coordinates
```
[the remark this answers is not in the document]
Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: Done.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 10]
```diff
@@ -65,19 +70,73 @@ public boolean isLastLeg(RouteProgress routeProgress) {
...
+    int remainingWaypointsCount = routeProgress.remainingWaypointsCount();
+    if (coordinatesSize < remainingWaypointsCount) {
       return null;
     }
-    List<Point> remainingCoordinates = coordinates.subList(coordinatesSize - remainingWaypoints, coordinatesSize);
+    String waypointIndices = routeOptions.waypointIndices();
+    if (waypointIndices == null) {
+      return null;
+    }
+    String[] wayPointIndices = waypointIndices.split(SEMICOLON);
```
[the remark this answers is not in the document]
Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: Yes, of course, I just missed it) Thanks for vigilance!

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 11]
```diff
@@ -65,19 +70,73 @@ public boolean isLastLeg(RouteProgress routeProgress) {
...
+    String[] remainingWaypointIndices = Arrays.copyOfRange(wayPointIndices,
+      wayPointIndices.length - remainingWaypointsCount, wayPointIndices.length);
+    List<Point> remainingCoordinates = new ArrayList<>();
+    try {
+      int firstRemainingWaypointIndex = Integer.valueOf(remainingWaypointIndices[FIRST_POSITION]);
+      remainingCoordinates = coordinates.subList(firstRemainingWaypointIndex, coordinatesSize);
+    } catch (NumberFormatException ex) {
+      Timber.e("Fail to convert waypoint index to integer");
```
[the remark this answers is not in the document]
Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: Done

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java [hunk 12]
```diff
@@ -65,19 +70,73 @@ public boolean isLastLeg(RouteProgress routeProgress) {
...
+    try {
+      int firstRemainingWaypointIndex = Integer.valueOf(remainingWaypointIndices[FIRST_POSITION]);
+      int traveledCoordinatesCount = firstRemainingWaypointIndex - ORIGIN_WAYPOINT_INDEX_THRESHOLD;
+      resultWaypointIndices = new Integer[remainingWaypointIndices.length + ORIGIN_WAYPOINT_NAME_THRESHOLD];
+      resultWaypointIndices[ORIGIN_WAYPOINT_INDEX] = Integer.valueOf(allWaypointIndices[ORIGIN_WAYPOINT_INDEX]);
+      for (int i = 0; i < remainingWaypointIndices.length; i++) {
+        resultWaypointIndices[i + 1] = Integer.valueOf(remainingWaypointIndices[i]) - traveledCoordinatesCount;
+      }
+    } catch (NumberFormatException ex) {
+      Timber.e("Fail to convert waypoint index to integer");
```
[the remark this answers is not in the document]
Comment (Lebedinsky) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/RouteUtils.java]: Done

File: libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/utils/RouteUtilsTest.java [hunk 13]
```diff
@@ -119,34 +124,106 @@ public void findCurrentBannerInstructions_adjustedDistanceRemainingRemovesCorrec
...
 
     List<Point> remainingWaypoints = routeUtils.calculateRemainingWaypoints(routeProgress);
 
-    assertEquals(2, remainingWaypoints.size());
-    assertEquals(Point.fromLngLat(7.890, 1.234), remainingWaypoints.get(0));
-    assertEquals(Point.fromLngLat(5.678, 9.012), remainingWaypoints.get(1));
+    Point[] expectedRemainingWaypoints = {Point.fromLngLat(4.56789, 0.12345)};
+
+    assertNotNull(remainingWaypoints);
+    assertArrayEquals(expectedRemainingWaypoints, remainingWaypoints.toArray());
```
[the remark this answers is not in the document]
Comment (Lebedinsky) [reply libandroid-navigation/src/test/java/com/mapbox/services/android/navigation/v5/utils/RouteUtilsTest.java]: Done

**Comment by Lebedinsky** [comment]
We decide to leave nullable types for calculation methods for now, cause for adding `NotNull` types we need to force MAS updates

**Pull request merged**